### PR TITLE
Fingerprint suppressions

### PR DESCRIPTION
**Pull request by monojack**

### CODE REVIEW

File: packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    this.suppressions = s;
+    return Promise.resolve();
+  }
+
+  suppress(problem: ValidationResult): Suppression[] | Promise<Suppression[]> {
+    const suppression = this.suppressions.find(s => {
+      return s.fingerprint && s.kind === this.kind && s.fingerprint === problem.fingerprints?.['monokleHash/v1'];
+    });
+
+    suppression && problem.suppressions?.push(suppression);
```
Comment (WitoDelnat) [inline packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts:19]: I think the suppression engine already handles this for you. I tried to keep the suppressors pure functions.

I also had to read this expression twice to understand what it does. A simple if-statement should improve readability:
```
if (suppression) {
  problem.suppressions.push(suppression);
}
```

File: packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  suppressions: FingerprintSuppression[] = [];
+
+  preload(s: FingerprintSuppression[]): Promise<void> {
+    this.suppressions = s;
+    return Promise.resolve();
+  }
+
+  suppress(problem: ValidationResult): Suppression[] | Promise<Suppression[]> {
+    const suppression = this.suppressions.find(s => {
```
Comment (WitoDelnat) [inline packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts:15]: This has O(n*m) complexity with unbound lists. I'd slightly optimise this given how easy it is:

```
preload() {
  const filteredSuppressions = suppressions.filter(s => s.fingerprint && s.kind === this.kind);
  this.suppressionsByFingerprint = keyBy(filteredSuppressions, fingerprint); // as Map
}

suppress() {
  const fingerprint = problem.fingerprints?.['monokleHash/v1'];
  if (!fingerprint) return [];
  const suppression = this.suppressionByFingerprint.get(fingerprint);
  // ...
}
```

This is once O(n) and then O(m)

Comment (monojack) [reply packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts:15]: Nice! Challenge accepted!

File: packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts (hunk 1 above)
Comment (monojack) [reply packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts:19]: Nice catch, you're right, the engine already handles that, I kinda just "copied" the redux-suppressor and adapted it but didn't really check what the engine does.

File: packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import {Suppression, SuppressionKind, ValidationResult} from '../../../common/sarif.js';
+import {FingerprintSuppression, Suppressor} from '../types.js';
+
+export class FingerprintSuppressor implements Suppressor<FingerprintSuppression> {
+  kind: SuppressionKind = 'external';
+  suppressions: FingerprintSuppression[] = [];
+  suppressionsMap: Map<FingerprintSuppression['fingerprint'], FingerprintSuppression> = new Map();
+
+  preload(suppressions: FingerprintSuppression[]): Promise<void> {
+    this.suppressions = suppressions;
```
Comment (WitoDelnat) [inline packages/validation/src/sarif/suppressions/plugins/FingerprintSuppressor.ts:10]: Nitpick: this is unused

**Review by WitoDelnat — APPROVED** [review]
LGTM

**Pull request merged**